core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/list/kopt/KOptListMove.java

```diff
@@ -0,0 +1,208 @@
+            int postShiftAmount) {
+        this.listVariableDescriptor = listVariableDescriptor;
+        this.entity = entity;
+        this.descriptor = descriptor;
+        this.equivalent2Opts = equivalent2Opts;
+        this.postShiftAmount = postShiftAmount;
+    }
+
+    @Override
+    protected AbstractMove<Solution_> createUndoMove(ScoreDirector<Solution_> scoreDirector) {
+        if (equivalent2Opts.isEmpty()) {
+            return this;
+        } else {
+            List<FlipSublistMove<Solution_>> inverse2Opts = new ArrayList<>(equivalent2Opts.size());
+            for (int i = equivalent2Opts.size() - 1; i >= 0; i--) {
+                inverse2Opts.add(equivalent2Opts.get(i).createUndoMove(scoreDirector));
+            }
+            return new UndoKOptListMove<>(listVariableDescriptor, entity, descriptor, inverse2Opts, -postShiftAmount);
+        }
+    }
+
+    @Override
+    protected void doMoveOnGenuineVariables(ScoreDirector<Solution_> scoreDirector) {
+        equivalent2Opts.forEach(move -> move.doMoveOnGenuineVariables(scoreDirector));
+        rotateToOriginalPositions(scoreDirector, listVariableDescriptor, entity, postShiftAmount);
+    }
+
+    private static <Solution_> void rotateToOriginalPositions(ScoreDirector<Solution_> scoreDirector,
+            ListVariableDescriptor<Solution_> listVariableDescriptor,
+            Object entity, int shiftAmount) {
+        InnerScoreDirector<Solution_, ?> innerScoreDirector = (InnerScoreDirector<Solution_, ?>) scoreDirector;
+        List<Object> listVariable = listVariableDescriptor.getListVariable(entity);
+
+        innerScoreDirector.beforeListVariableChanged(listVariableDescriptor, entity, 0, listVariable.size());
+
+        Collections.rotate(listVariable, shiftAmount);
+
+        innerScoreDirector.afterListVariableChanged(listVariableDescriptor, entity, 0, listVariable.size());
+    }
+
+    @Override
+    public boolean isMoveDoable(ScoreDirector<Solution_> scoreDirector) {
+        return !equivalent2Opts.isEmpty();
+    }
+
+    @Override
+    public Move<Solution_> rebase(ScoreDirector<Solution_> destinationScoreDirector) {
+        List<FlipSublistMove<Solution_>> rebasedEquivalent2Opts = new ArrayList<>(equivalent2Opts.size());
+        for (FlipSublistMove<Solution_> twoOpt : equivalent2Opts) {
+            rebasedEquivalent2Opts.add(twoOpt.rebase(destinationScoreDirector));
+        }
+        return new KOptListMove<>(listVariableDescriptor, destinationScoreDirector.lookUpWorkingObject(entity),
+                descriptor, rebasedEquivalent2Opts, postShiftAmount);
+    }
+
+    @Override
+    public String getSimpleMoveTypeDescription() {
+        return descriptor.getK() + "-opt(" + listVariableDescriptor.getSimpleEntityAndVariableName() + ")";
+    }
+
+    @Override
+    public Collection<?> getPlanningEntities() {
+        return List.of(equivalent2Opts.get(0).getEntity());
+    }
+
+    @Override
+    public Collection<?> getPlanningValues() {
```

Comment:
This is used in tabu search, therefore on the hot path.
I'd avoid the stream here, let's convert to a normal loop.